_posts/2022-11-20-bjarne_gsoc.md

```diff
@@ -0,0 +1,104 @@
+and pyedr] (PRs [#42] and [#50]). Both packages read EDR files, but one returns the
+data as a pandas DataFrame, the other as a dictionary of NumPy arrays. Both also
+expose a function to return a dictionary of units of the energy terms found in the file ([PR #56]).
+
+Example:
+
+```python
+import pyedr
+file = "path/to/edr/file.edr"
+energy_dictionary = pyedr.edr_to_dict(file)
+unit_dictionary = pyedr.get_unit_dictionary(file)
+```
+
+
+## EDRReader
+With Pyedr available, I started work on the implementation of an EDRReader in MDAnalyis ([PR #3749]).
+Here, I benefited hugely from the existing [AuxReader framework].
+However, while working on the reader, it quickly became apparent that the auxiliary API would need to be changed to accommodate
+the large number of terms found in EDR files. While the [XVGReader] still works as previously, the new base case for adding auxiliary data assumes a dictionary to be passed. The dictionary maps the name to be used in MDAnalysis to the names read from the EDR file. This is shown in the following minimal working example:
+
+```python
+import MDAnalysis as mda
+from MDAnalysisTests.datafiles import AUX_EDR, AUX_EDR_TPR, AUX_EDR_XTC
+term_dict = {"temp": "Temperature", "epot": "Potential"}
+aux = mda.auxiliary.EDR.EDRReader(AUX_EDR)
+u = mda.Universe(AUX_EDR_TPR, AUX_EDR_XTC)
+u.trajectory.add_auxiliary(term_dict, aux)
+```
+
+Aside from this API change, the EDRReader can do everything the XVGReader can. In addition to that, it
+has some new functionality.
+* Because EDR files can become reasonably large, a memory warning will be issued when more than a gigabyte of storage is used by the auxiliary data. This default value of 1 GB can be changed by passing a value as `memory_limit` when creating the EDRReader object.
+* EDR files store data of a large number of different quantities, so it is important to know their units as well. The EDRReader therefore has a `unit_dict` attribute that contains this information. By default, units found in the EDR file will be converted to [MDAnalysis base units] on reading. This can be disabled by setting `convert_units` to False on creation of the reader.
+* In addition to associating data with trajectories, the EDRReader can also return the NumPy arrays of selected data, which is useful for plotting, for example. This is done via the EDRReader's `get_data` method.
+
+Additionally, the new auxiliary readers allow the selection of frames based on the
+values of the auxiliary data. For example, it is possible to select only frames
+with a potential energy below a certain threshold as follows:
+```python
+u = mda.Universe(AUX_EDR_TPR, AUX_EDR_XTC)
+term_dict = {"epot": "Potential"}
+u.trajectory.add_auxiliary(term_dict, aux)
+selected_frames = np.array([ts.frame for ts in u.trajectory if ts.aux.epot < -524600])
+```
+More details on the EDRReader's functionality can be found in the [MDAnalysis User Guide].
+
```

Comment:
Would you be able to comment on writing a new AuxReader? What is the outlook for a new developer coming in?

Reply:
You enhanced the API and made it more general so that it should become easier to accommodate other energy file formats such as ... (AMBER, NAMD, generic CSV, numpy arrays) — simply go back to your proposal. Say what would need to be done, what are the challenges, which of your additions make it easier?

Doesn't have to be long, just one paragraph, but it shows that you've been developing a framework for the future.